src/sqlfluff/rules/structure/ST10.py

```diff
@@ -84,13 +101,18 @@ def _eval_gen(self, context: RuleContext) -> Iterator[LintResult]:
                     # Should be unreachable with correctly parsed tree
                     continue  # pragma: no cover
 
+                if lhs.is_templated or rhs.is_templated:
+                    continue
+
                 # literals need explicit handling (due to well-defined allow-list)
                 if lhs.is_type("literal") and rhs.is_type("literal"):
                     expr_s = f"{lhs.raw_normalized()} {seg.raw} {rhs.raw_normalized()}"
                     if expr_s in allowable_literal_expressions:
                         # ignore based on allowlist
                         continue
                 else:
+                    if lhs.type != rhs.type:
+                        continue
```

Comment:
@keraion This effectively fixes your example and makes the overall checker somewhat stricter, which I assume is a good thing.

Reply:
Yeah, I like this.